src/FindIntrinsics.cpp

```diff
@@ -0,0 +1,714 @@
+        return true;
+    }
+    return false;
+}
+
+bool is_safe_for_add(const Expr &e) {
+    return is_safe_for_add(e, e.type().bits() / 2 - 1);
+}
+
+// We want to find and remove an add of 'round' from e.
+Expr find_and_subtract(const Expr &e, const Expr &round) {
+    if (const Add *add = e.as<Add>()) {
+        Expr a = find_and_subtract(add->a, round);
+        if (!a.same_as(add->a)) {
+            return Add::make(a, add->b);
+        }
+        Expr b = find_and_subtract(add->b, round);
+        if (!b.same_as(add->b)) {
+            return Add::make(add->a, b);
+        }
+    } else if (const Sub *sub = e.as<Sub>()) {
+        Expr a = find_and_subtract(sub->a, round);
+        if (!a.same_as(sub->a)) {
+            return Sub::make(a, sub->b);
+        }
+        // We can't recurse into the negatve part of a subtract.
+    } else if (can_prove(e == round)) {
+        return make_zero(e.type());
+    }
+    return e;
+}
+
+Expr to_rounding_shift(const Call *c) {
+    if (c->is_intrinsic(Call::shift_left) || c->is_intrinsic(Call::shift_right)) {
+        internal_assert(c->args.size() == 2);
+        Expr a = c->args[0];
+        Expr b = c->args[1];
+
+        // Helper to make the appropriate shift.
+        auto rounding_shift = [&](const Expr &a, const Expr &b) {
+            if (c->is_intrinsic(Call::shift_right)) {
+                return rounding_shift_right(a, b);
+            } else {
+                return rounding_shift_left(a, b);
+            }
+        };
+
+        // The rounding offset for the shift we have.
+        Type round_type = a.type().with_lanes(1);
+        if (Call::as_intrinsic(a, {Call::widening_add})) {
+            round_type = round_type.narrow();
+        }
+        Expr round;
+        if (c->is_intrinsic(Call::shift_right)) {
+            round = simplify((make_one(round_type) << max(cast(b.type().with_bits(round_type.bits()), b), 0)) / 2);
+        } else {
+            round = simplify((make_one(round_type) >> min(cast(b.type().with_bits(round_type.bits()), b), 0)) / 2);
+        }
+
+        // We can always handle widening adds.
+        if (const Call *add = Call::as_intrinsic(a, {Call::widening_add})) {
+            if (can_prove(lower_intrinsics(add->args[0]) == round)) {
+                return rounding_shift(cast(add->type, add->args[1]), b);
+            } else if (can_prove(lower_intrinsics(add->args[1]) == round)) {
+                return rounding_shift(cast(add->type, add->args[0]), b);
+            }
+        }
+
+        // If it wasn't a widening or saturating add, we might still
+        // be able to safely accept the rounding.
+        Expr a_less_round = find_and_subtract(a, round);
+        if (!a_less_round.same_as(a)) {
+            // We found and removed the rounding. However, we may have just changed
+            // behavior due to overflow. This is still save if the type is not
+            // overflowing, or we can find a widening add or subtract in the tree
```

Comment:
Why does finding the round term make it safe? What if the round term caused overflow in the original and you just removed it?